### DISCUSSION TITLE
SAWarning: Object of type <MyModel> not in session, add operation along 'RelatedModel.my_model' won't proceed

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#8330] [Usage Questions]
I have two models 

```python

class MyModel(Base):
    __tablename__ = 'my_model'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    rel_model = relationship("RelatedModel", back_populates='my_model')

class RelatedModel(Base):
    __tablename__ = 'related_model'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    my_model_id = Column (Integer,   ForeignKey('my_model.id'))
```
I create an instance of `MyModel` and I assign to the relationship field an instance of `RelatedModel` :
```python
instance = MyModel(name="blabla")
related_instance = RelatedModel(name="related_blabla")
instance.rel_model = related_instance
```
I add it to the session, and for some reason (if something fail) I want to expunge it from the session :
```python
try:
    session.add(instance)
    raise Exception()
except Exception:
    session.expunge(instance)
```
everything is working as expected. However, after this expunge, if I select `MyModel` from the session again : 

```python
query = select(MyModel).outerjoin(RelatedModel).where(RelatedModel.id.is_(None))
instances = session.execute(query).unique().scalars().all()
``` 
I got this warning : 
```python
SAWarning: Object of type <MyModel> not in session, add operation along 'RelatedModel.my_model' won't proceed
```
Did I do something wrong ? what does this warning exactly mean?

**Comment by zzzeek** [comment]
well you expunged the object from the Session, yet it's still present in related_model.my_model.  That is, one object is in the session, the other object it points to is not.   so it's pointing out that this object is not going to be persisted or anything like that when you flush.

the solution is to not use session.expunge(), there's no need to do that pattern you are doing.  if add() fails, the object will not be in the session.

**Reply by jossefaz** [reply]
thanks for this explanation.
It does explain me the issue. however, I did not reproduce the code exactly as it is. since the try/except does really apply on the session add. session add occurs before. 
it is more something like :  
```python
session.add(instance)
try:
    # some calculation process that could fail
    raise Exception()
except Exception:
    session.expunge(instance)
```
How can I solve this then ?

**Reply by zzzeek** [reply]
why do you need to expunge the instance ?

**Reply by jossefaz** [reply]
Okay, the "real" scneario is as follow : 
I retrieve from DB potential rows, that need to be updated. Each rows is mapped to ORM instances
I run a for loop on those instances, and pass each instance through a kind of "update pipeline". 
At the end of the for loop : I commit.
If one of the update step fails on an instance, I just want to revert all the changes that was made by previous update steps **on that particular instance only** (so I cannot use session.rollback, since it would rollback all other instances as well....)

So I was looking on a way to rollback changes and not commit a particular instances, among all the instances that are getting updated in this for loop.

In this example, I used `add` because it was a unit test...

**Reply by zzzeek** [reply]
OK for this case use SAVEPOINT, which in the Session is using the begin_nested() method.   Here is a doc: https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#using-savepoint

this use case needs more docs in general since people tend not to know about this.

**Reply by zzzeek** [reply]
also that last paragraph looks wrong...let me adjust

**Reply by jossefaz** [reply]
That is exactly what I was looking for ! thank you soo much
